File: Serenity.Script.UI/Misc/Recaptcha.cs

namespace Serenity
{
    using jQueryApi;
    using System;
    using System.Collections;
    using System.Collections.Generic;
    using System.Html;
    using System.Runtime.CompilerServices;

    [Editor, Element("<div />")]
    public class Recaptcha : Widget<RecaptchaOptions>, IStringValue
    {
        public Recaptcha(jQueryObject div, RecaptchaOptions opt)
            : base(div, opt)
        {
            this.element.AddClass("g-recaptcha")
                .Attribute("data-sitekey", options.SiteKey);

            if (Window.Instance.As<dynamic>().grecaptcha == null &&
                J("script#RecaptchaInclude").Length == 0)
            {
                var src = "https://www.google.com/recaptcha/api.js";
                src += "?hl=" + (options.Language ?? J("html").GetAttribute("lang") ?? "");

                J("<script/>").Attribute("id", "RecaptchaInclude").Attribute("src", src)
                    .AppendTo(Document.Body);
            }

            var input = J("<input />").InsertBefore(this.element).Attribute("id", this.UniqueName + "_validate")
                .Value("x").CSS(new JsDictionary<string, object>
                {
                    { "visibility", "hidden" },
                    { "width", "0px" },
                    { "height", "0px" },
                    { "padding", "0px" }
                });

            var self = this;
            input.AddValidationRule(this.uniqueName, e =>
            {
                if (string.IsNullOrEmpty(this.Value))
                    return Q.Text("Validation.Required");

                return null;
            });
        }

        public string Value
        {
            get
            {
                return this.element.Find(".g-recaptcha-response").GetValue();
            }
            set
            {
                // ignore
            }
        }
    }

    [Imported, Serializable]
    public class RecaptchaOptions
    {
        public string SiteKey { get; set; }
        public string Language { get; set; }
    }
}

File: src/Serenity.Scripts/CoreLib/UI/Editors/Recaptcha.ts
namespace Serenity {

    export interface RecaptchaOptions {
        siteKey?: string;
        language?: string;
    }

    @Decorators.registerEditor('Serenity.Recaptcha', [IStringValue])
    @Decorators.element("<div/>")
    export class Recaptcha extends Widget<RecaptchaOptions> implements IStringValue {
        constructor(div: JQuery, opt: RecaptchaOptions) {
            super(div, opt);

            this.element.addClass('g-recaptcha').attr('data-sitekey', this.options.siteKey);
            if (!!(window['grecaptcha'] == null && $('script#RecaptchaInclude').length === 0)) {
                var src = 'https://www.google.com/recaptcha/api.js';
                var lng = this.options.language;
                if (lng == null) {
                    lng = Q.coalesce($('html').attr('lang'), '');
                }
                src += '?hl=' + lng;
                $('<script/>').attr('id', 'RecaptchaInclude').attr('src', src).appendTo(document.body);
            }

            var valInput = $('<input />').insertBefore(this.element)
                .attr('id', this.uniqueName + '_validate').val('x');

            var gro = {};
            gro['visibility'] = 'hidden';
            gro['width'] = '0px';
            gro['height'] = '0px';
            gro['padding'] = '0px';

            var input = valInput.css(gro);
            var self = this;

            Q.addValidationRule(input, this.uniqueName, e => {
                if (Q.isEmptyOrNull(this.get_value())) {
                    return Q.text('Validation.Required');
                }
                return null;
            });
        }

        get_value(): string {
            return this.element.find('.g-recaptcha-response').val();
        }

        set_value(value: string): void {
            // ignore
        }
    }
}